lib/escalate.rb

```diff
@@ -81,8 +83,20 @@ def default_escalate_logger
 
     # Registers an escalation callback to be executed when `escalate`
     # is invoked.
-    def on_escalate(&block)
-      on_escalate_blocks.add(block)
+    #
+    # @param [boolean] log_first: true
+    #   whether escalate should log first before escalating, or leave the logging to the escalate block
+    def on_escalate(log_first: true, &block)
+      if log_first
```

Comment:
Does this need to be split into two separate groups? Can we simplify this by having a private class variable defined `disable_logging` which starts as `false` and then when `log_first` comes in as `false` it flips to `true`? I think overall this would reduce the complexity of holding two sets of callbacks.

```
def on_escalate(log_first: true, &block)
  disable_logging = self.disable_logging || !log_first
  on_escalate_blocks.add(block)
end
```

Reply:
I didn't want to be stateful about the order of callback declaration.

Reply:
What I really wanted was to store an attribute on the callbacks. Turns out that Ruby actually allows this. What do you think?
```
block.instance_variable_set(:@log_first, log_first)
```
Then later I could check those:
```
on_escalate_blocks.any? { |block| block.instance_variable_get(:@log_first) }
```

Reply:
@jebentier I took a shot at it here: https://github.com/Invoca/escalate/pull/2/commits/2e946ae65f5802d588e5d022bcd811a0ade42fd9

Seems much cleaner to me. It's slightly thin ice to put instance variables on blocks since Ruby doesn't document that as a feature. But these unit tests will notice if they ever retract that. Seem ok?

Reply:
> I didn't want to be stateful about the order of callback declaration.

I get what you're talking about now with this comment now after seeing this new implementation.

> Seems much cleaner to me. It's slightly thin ice to put instance variables on blocks since Ruby doesn't document that as a feature. But these unit tests will notice if they ever retract that. Seem ok?

Definitely agree that this seems much cleaner. I'm good with this thin contract. We can always think about different approaches later as well if we are concerned with the thin nature of this practice.

Reply:
Cool. I feel pretty good with the "thin ice" here since it's completely encapsulated inside this gem.
Hmm, I wonder if it'll work in JRuby?!

Reply:
🤷